### PR TITLE
Bugfixing #1355

### LINKED ISSUE #1543: Expression for default value evalues in opposite way
In the project input-testing/forms_derived there is field "Name" on tab "Data" with default value set by expression "if("Danger" IS 'false' ,'Bear safe','Bear dangerous')". 
Please note that to see Data tab you need to insert some value into "Information" tab into field "Value"

So it should set name "Bear safe" when Danger = FALSE and "Bear dangerous" then Danger = TRUE, but it do just the opposite.

### PR DESCRIPTION
**Pull request by sklencar**
From #1533 :
- [x] 2. 'Reuse last value' feature is not working - fixed in https://github.com/lutraconsulting/input/pull/1549
- [x] 3. After restarting the app, layer which is not present in the them, can be set as default for recording. This can be very confusing.
- [x] 7. issue described in ticket Multiline widget text disappears
	-> used onTextChanged for text widget
- [x] 8. closes #1543

### CODE REVIEW

**Comment by jozef-budac** [comment]
There is still problem with `Reuse last value` feature. When this feature is turn on, in 2nd recorded feature, stored values are used in new form, but checkbox are missing.

https://user-images.githubusercontent.com/83582181/126116009-ffd90345-6aa1-4d4d-ae97-fb60046f00ec.mp4

**Comment by jozef-budac** [comment]
There is still problem with 5. point
After I choose another Car manufacturer with search (or without), `Car model` should be cleared, but is isn't. 
When I store the feature, Car model is not stored (which is correct).

https://user-images.githubusercontent.com/83582181/126117306-23d333d0-397a-496c-949d-5a8a41830ffe.mp4

**Comment by jozef-budac** [comment]
7. I have still the same problem with this PR


https://user-images.githubusercontent.com/83582181/126117965-98b9d69c-6641-4fa9-9915-09b0c8e0b301.mp4

**Comment by tomasMizera** [comment]
> There is still problem with 5. point
> After I choose another Car manufacturer with search (or without), `Car model` should be cleared, but is isn't.
> When I store the feature, Car model is not stored (which is correct).
> 
>  20210719_090438.mp4

I think this one is fixed in #1545

**Comment by tomasMizera** [comment]
Can you please rebase master? There are several fixes merged, thanks

**Comment by sklencar** [comment]
PR rebased with master (some other bugfixes have been applied).

Issue 5. - It has been moved to separate issue https://github.com/lutraconsulting/input/issues/1559
Issue 7. - I cant replicate it. Can anyone confirm is still present?

File: app/loader.cpp [hunk 1]
```diff
@@ -107,7 +108,13 @@ bool Loader::forceLoad( const QString &filePath, bool force )
...
     mActiveLayer.resetActiveLayer();
     mMapThemeModel.reloadMapThemes( mProject );
-    setActiveLayerByName( mAppSettings.defaultLayer() );
+
+    QgsVectorLayer *defaultLayer = mRecordingLayerPM.layerFromLayerName( mAppSettings.defaultLayer() );
+    if ( defaultLayer )
+      setActiveLayer( defaultLayer );
+    else
+      setActiveLayer( mRecordingLayerPM.firstUsableLayer() );
+
```
Comment (tomasMizera) [inline app/loader.cpp:117]: Similar thing is happening here: https://github.com/lutraconsulting/input/blob/c8c8e23e584763ae36093f0ab33961a1ae628fbf/app/qml/main.qml#L211

Can we maybe erase one of these two?

File: app/attributes/attributecontroller.cpp [hunk 2]
```diff
@@ -729,12 +729,6 @@ void AttributeController::recalculateDerivedItems( bool isFormValueChange, bool
         }
       }
 
-      if ( !isFormValueChange ) // reset state for new feature layer pair
```
Comment (tomasMizera) [inline app/attributes/attributecontroller.cpp:732]: Have you checked if the state is reset and new features do not inherit it?

File: app/loader.cpp (hunk 1 above)
Comment (sklencar) [reply app/loader.cpp:117]: Similar, but I think a bit different. 
The former is called when a project is loaded and handles setting theme and layer accordingly.
The latter is used in `updateRecordToolbar` function which is called in several cases. Most relevant reason to keep it there is that it handles the case when:

1. user changes map theme which doesn't contain current active layer
2. starts to digitizing

Another option would be to replace/fix `Loader::setActiveLayerByName` used in `Loader::forceLoad`.
Essentially, the fix would be similar (checking if default layer is present in digitizing proxy layer model).

**Comment by sklencar** [comment]
I am afraid that issue related with calculating validity and evaluating constrains is broken. Seems to be complex issue and thus rough estimate to revise all the process is rather counted in days.

Known issues:
- invalid state of field can be set to valid easily - value out of range can be reset just by clicking on another field and back to the field. Most likely, editingFinished is triggered which triggers following actions. It can be replicated as Jozef described.
- invalid default value is ignored for new features - if default value breaks the range from config (config range 0-10, 11 is set as default value), the feature can be saved anyway

**Pull request merged**